README.md
# FIO PHP SDK

Content to be based on SDK READMEs found at:

- https://github.com/fioprotocol/fiosdk_kotlin
- https://github.com/fioprotocol/fiojs
- https://github.com/fioprotocol/fio-go
- https://github.com/fioprotocol/fiosdk_ios

## Installation in your project

Editing your composer.json file is so 2015. Use the command line tooling or your IDE! :-)

```shell
composer require clarkphp/fio-php-sdk
```

## Unit Tests and Code Quality

Run the tests as shown below. For more on the tests, generating code coverage
reports, and static analysis tools, see [CONTRIBUTING.md](CONTRIBUTING.md).

```shell
$ cd /path/to/fio-client/
$ phpunit
```

## Contributing

If you'd like to contribute, please see the [CONTRIBUTING.md](CONTRIBUTING.md) file.

## Versions

| Version | Status     | Packagist           | Namespace    | Repo                | Docs                | PSR-18 | PHP Version |
|---------|------------|---------------------|--------------|---------------------|-------------------|-------|-------------|
| 0.x     | Initial Dev | `clarkphp/fio-php-sdk` | `ClarkPhp\FioPhpSdk` | [v0][clark-fio-0-repo] | [v0][clark-fio-0-docs] | Yes | >= 7.3 |

[clark-fio-0-repo]: https://github.com/clarkphp/fio-php-sdk
[clark-fio-0-docs]: https://github.com/clarkphp/fio-php-sdk/tree/main/docs

CHANGELOG.md
# Changelog

This file documents, in reverse chronological order by release, all notable changes to this project.

- See https://phly.github.io/keep-a-changelog/ for a tool to help keep changelogs up to date.

## 0.0.2 - 2021-01-02
### Added
- Very initial handling of error response from the API

### Changed
- A lot, but, c'mon, this is nowhere near Alpha

### Deprecated
- Nothing

### Removed
- Nothing

### Fixed
- Nothing

## 0.0.1 - 2021-01-01
### Added
- Client can do basic FIO address availability and validity checks

### Changed
- Replace direct use of Guzzle with FIOClient in example script

### Deprecated
- Nothing

### Removed
- Nothing

### Fixed
- Nothing

## 0.0.0 - 2020-12-31
### Added
- Initial files only. No functionality whatsoever!

### Changed
- Nothing

### Deprecated
- Nothing

### Removed
- Nothing

### Fixed
- Nothing


tests/FioClientTest.php
<?php

/**
 * @see       https://github.com/clarkphp/fio-php-sdk for the canonical source repository
 * @copyright https://github.com/clarkphp/fio-php-sdk/blob/main/LICENSE.md
 * @license   https://github.com/clarkphp/fio-php-sdk/blob/main/LICENSE.md
 */

declare(strict_types=1);

namespace ClarkPhpTest\FioPhpSdk;

use ClarkPhp\FioPhpSdk\FioClient;
use GuzzleHttp\Client as GuzzleClient;
use PHPUnit\Framework\TestCase;

// for unit tests only, not needed by users of FioClient
use GuzzleHttp\Psr7\Request as GuzzleRequest;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Psr7\Utils;

class FioClientTest extends TestCase
{
    private const BASE_URI = 'http://testnet.fioprotocol.io/v1/chain/';
    private const HEADER_JSON_CONTENT_TYPE = ['content-type' => 'application/json'];

    /** @var FioClient */
    public $fioClient;

    /** @var GuzzleClient  */
    public $httpClient;

    public function setUp(): void
    {
        $this->httpClient = new GuzzleClient(
            [
                'base_uri' => self::BASE_URI,
                'headers'  => self::HEADER_JSON_CONTENT_TYPE
            ]
        );

        $this->fioClient = new FioClient($this->httpClient);
    }

    public function tearDown(): void
    {
        unset($this->fioClient);
    }

    /**
     * Please note that Network test group is by default not invoked, because accessing
     * remote resources is normally a test smell. See phpunit.xml.dist groups section.
     * Copy to phpunit.xml and change 'Network' to 'default' or something else to run
     * Network tests.
     *
     * @group Network
     * group choose-a-good-test-group-name
     */
    public function testIsFioNameAvailable()
    {
        $this->assertTrue($this->fioClient->isFioNameAvailable('idonotexist@idonotexist'));

        $this->assertFalse($this->fioClient->isFioNameAvailable('fiodemo@edgetest'));

        $this->expectExceptionMessage('has invalid format. See https://developers.fioprotocol.io/api/api-spec/models/fio-address');
        $this->fioClient->isFioNameAvailable('');
    }

    /** @group Network */
    public function testIsFioNameRegistered()
    {
        $this->assertTrue($this->fioClient->isFioNameRegistered('fiodemo@edgetest'));
    }

    /** @group Validation */
    public function testIsFioNameValid()
    {
        $this->assertFalse($this->fioClient->isFioNameValid('qa'));
        $this->assertTrue($this->fioClient->isFioNameValid('a@b'));

        $this->assertFalse($this->fioClient->isFioNameValid('no-trailing-dash-@domain'));
        $this->assertFalse($this->fioClient->isFioNameValid('name@no-trailing-dash-'));

        $this->assertFalse($this->fioClient->isFioNameValid('-no-leading-dash@domain'));
        $this->assertFalse($this->fioClient->isFioNameValid('name@-no-leading-dash'));

        $this->assertFalse($this->fioClient->isFioNameValid('no_underscore@domain'));
        $this->assertFalse($this->fioClient->isFioNameValid('name@no_underscore'));

        $this->assertTrue($this->fioClient->isFioNameValid('64-chars-is-not-too-long-to-make-for-valid-fio-address@fiodomain'));
        $this->assertFalse($this->fioClient->isFioNameValid('65-chars-is-too-long-to-be-a-valid-fio-address@listen-to-this-bud'));

        $this->assertTrue($this->fioClient->isFioNameValid('1-numbers-are-okay-1@domain1'));
        $this->assertTrue($this->fioClient->isFioNameValid('vAlId-foRmaT1@FIO-address1'));
    }

    /**
     * @group Network
     * @group ErrorHandling
     */
    public function testExtractError404Details()
    {
        $message = $this->callNonExistentEndPoint();
        $this->assertSame(404, $this->fioClient->getResponseErrorNumber());
        $this->assertSame('Not Found', $this->fioClient->getResponseErrorMessage());
        $this->assertSame($message, 'Unknown Endpoint');
    }

    private function callNonExistentEndPoint()
    {
        try {
            $this->httpClient->send(
                (new GuzzleRequest('POST', 'idonotexist'))
                    ->withBody(Utils::streamFor('{"fio_name":"wilecoyote"}'))
            );
        } catch(RequestException $e) {
            return $this->fioClient->extractErrorDetailsFromRequest($e);
        }
    }
}

src/FioClient.php
<?php

/**
 * @see       https://github.com/clarkphp/fio-php-sdk for the canonical source repository
 * @copyright https://github.com/clarkphp/fio-php-sdk/blob/main/LICENSE.md
 * @license   https://github.com/clarkphp/fio-php-sdk/blob/main/LICENSE.md
 */

declare(strict_types=1);

namespace ClarkPhp\FioPhpSdk;

use Exception;
use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Psr7\Request as GuzzleRequest;
use GuzzleHttp\Psr7\Utils;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\StreamInterface;

final class FioClient
{
    // @todo Use factories to allow use of any PSR-18-compliant client, removing dependency on particular implementations

    const FIO_ADDR_VALID_FORMAT = '/^(?:(?=.{3,64}$)[a-z\d]{1}(?:(?!-{2,}))[a-z\d-]*(?:(?<!-))@[a-z\d]{1}(?:(?!-{2,}))[a-z\d-]*(?:(?<!-))$)/i';

    /** @var GuzzleClient */
    private $httpClient;

    /** @var ResponseInterface|null */
    private $errorResponse;

    /** @var string */
    private $errorDetails;

    /** @var \stdClass|null */
    private $responseBody;

    public function __construct(GuzzleClient $client)
    {
        $this->httpClient = $client;
    }

    public function isFioNameRegistered(string $nameToCheck): bool
    {
        return ! $this->isFioNameAvailable($nameToCheck);
    }

    public function isFioNameAvailable(string $nameToCheck): ?bool
    {
        $this->errorResponse = null;

        if (!$this->isFioNameValid($nameToCheck)) {
            throw new Exception(
                "'$nameToCheck' has invalid format. See https://developers.fioprotocol.io/api/api-spec/models/fio-address"
            );
        }

        try {
            return '{"is_registered":0}' === (string)(
                $this->httpClient->send(
                    (new GuzzleRequest('POST', 'avail_check'))
                        ->withBody(Utils::streamFor('{"fio_name":"' . $nameToCheck . '"}'))
                ))->getBody();
        } catch (RequestException $e) {
                $this->errorDetails = $this->extractErrorDetailsFromRequest($e);
        }

        return null;
    }

    public function isFioNameValid(string $name): bool
    {
        return (bool)preg_match(self::FIO_ADDR_VALID_FORMAT, $name);
    }

    //@todo Refactor error handling to a separate class
    // should this return an error object of some sort?
    public function extractErrorDetailsFromRequest(RequestException $e): string
    {
        if ($e->hasResponse()) {
            $this->errorResponse = $e->getResponse();
            $this->responseBody = $this->convertStreamToObject($e->getResponse()->getBody());
            if (0 !== ($errNum = $e->getResponse()->getStatusCode())) {
                $methodName = "extractError{$errNum}Details";
                return $this->$methodName($this->responseBody);
            }
        }
        return '';
    }

    // @todo possibly return a "useful" PHP class instead?
    private function convertStreamToObject(StreamInterface $body): \stdClass
    {
        return json_decode((string) $body);
    }

    public function getResponseErrorNumber(): int
    {
        return $this->errorResponse->getStatusCode();
    }

    public function getResponseErrorMessage(): string
    {
        return $this->errorResponse->getReasonPhrase();
    }

    /**
     * @see https://developers.fioprotocol.io/api/api-spec/models/error-400
     * @return object
     */
    private function extractError400Details(): object
    {
    /*
    { documentation page
      "type": "object",
      "title": "Error 400",
      "required": [
        "type",
        "message",
        "fields"
      ],
      "properties": {
        "type": {
          "type": "string",
          "default": "invalid_input",
          "description": "invalid_input"
        },
        "message": {
          "type": "string",
          "default": "An invalid request was sent in, please check the nested errors for details.",
          "description": "An invalid request was sent in, please check the nested errors for details."
        },
        "fields": {
          "type": "array",
          "items": {
            "type": "object",
            "properties": {
              "name": {
                "type": "string",
                "description": "Name of the field which triggered the error"
              },
              "value": {
                "type": "string",
                "description": "Value which was sent in and which triggered the error"
              },
              "error": {
                "type": "string",
                "description": "Error message"
              }
            }
          }
        }
      }
    }

     */

        return new \stdClass();
    }

    /**
     * @see https://developers.fioprotocol.io/api/api-spec/models/error-403
     * @return object
     */
    private function extractError403Details():object
    {
        /*
    {
      "type": "object",
      "title": "Error 403",
      "properties": {
        "type": {
          "type": "string",
          "description": "Type of 403 error",
          "example": "invalid_signature"
        },
        "message": {
          "type": "string",
          "example": "Request signature not valid or not allowed.",
          "description": "Message associated with the 403 type"
        }
      },
      "required": [
        "type",
        "message"
      ]
    }
        */
        return new \stdClass();
    }

    /**
     * @see https://developers.fioprotocol.io/api/api-spec/models/error-404
     *
     * @param   \stdClass  $apiResponseBody
     * @return string
     */
    private function extractError404Details(\stdClass $apiResponseBody): string
    {
        /* documentation page
        {
          "type": "object",
          "title": "Error 404",
          "properties": {
            "message": {
              "type": "string",
              "description": "Call specific error message.",
              "example": "Public address not found"
            }
          }
        }

        Value from actual run:
        stdClass Object
        (
            [code] => 404
            [message] => Not Found
            [error] => stdClass Object
                (
                    [code] => 0
                    [name] => exception
                    [what] => unspecified
                    [details] => Array
                        (
                            [0] => stdClass Object
                                (
                                    [message] => Unknown Endpoint
                                    [file] => http_plugin.cpp
                                    [line_number] => 353
                                    [method] => handle_http_request
                                )
                        )
                )
        )
        */

        return (($apiResponseBody->error->details)[0])->message ?? '';
    }
}